### PR TITLE
Time series based workload desc order optimization through reverse segment read

### PR DESCRIPTION
**Pull request by gashutos**
### Description

`EDIT -> After incorporating comments/suggestions from Nick, Andriya & Andorss, we decided to limit this change only to @timestamp field indices and also we changed default segment search order for these indices as Descending (from latest to the oldest segments), and also safe guarded rare ASC queries on time series based workload not to traverse on default DESC order to avoid any regression. The perf gains still remains similar.`

As described in issue [OpenSearch-6814](https://github.com/opensearch-project/OpenSearch/issues/6814), desc order sort queries are taking more time (3x or more) compare to asc order queries on same data set (time series based).
That is because all latest documents in time series workload are in latest segments when our IndexSearcher search documents in order of oldest segments to newest segments. 

In this PR, we changed the leafReaderContext (segments) order in ContextIndexSearcher in reverse if query type is descending order sort. This will by default behaviour for all type of workloads and not just for time series based workloads. However it will only benefit to workload which has ever increasing data values. And for other type of workload, it wont regress or impact since data is randomly distributed across all segments.
But as a precaution, in case any user/workload gets search latency hit, I have added index level setting to disable this optimization.

### Optimization gains

We are seeing 3x to 5x gains on below configuration.

- On http_logs workload with disabled force merge, so each shard has around 100-150  segments.
- instance type, r6g.2xlarge.
- units are in ms.

| Sort Type      | hits = 5 | hits = 100 | hits = 1000 |
| ----------- | ----------- | -------- | -------- |
| Desc   | 150        | 462 | 621 |
| Desc (with this change)   | 49        | 112| 125|

### Issues Resolved
[OpenSearch-6814](https://github.com/opensearch-project/OpenSearch/issues/6814)

### Next action item
On same segment of time series based workload, ASC queiry is faster compare to descending order. We need to look further to that.

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by gashutos** [comment]
@nknize @reta , check once you get time, if this appraoch looks good. The optimization gain is pretty good.

File: server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java:274]: Is it possible iterate on the existing list in reverse order based on the boolean flag? That will prevent the additional allocation and reverse operation.

File: server/src/main/java/org/opensearch/search/SearchService.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/search/SearchService.java:238]: Can this be more helpful as an index setting instead? A user can choose to apply this for the indices they deem useful within a cluster.

File: server/src/main/java/org/opensearch/search/SearchService.java
Comment (mgodwan) [inline server/src/main/java/org/opensearch/search/SearchService.java:1102]: Could we cover this logic in a UT in SearchServiceTests?

**Review by mgodwan — COMMENTED** [review]
Is it possible to add a test to SimpleSortIT for the setting the optimization as true for the node/index and verifying that the results are same irrespective of the leafReaderContext iteration order?

**Comment by gashutos** [comment]
> Is it possible to add a test to SimpleSortIT for the setting the optimization as true for the node/index and verifying that the results are same irrespective of the leafReaderContext iteration order?

Thanks @mgodwan for the review. Yes I added random flag for integ test, which will randomly enable/disble this optimization flag while creating index, so few tests will run with/without this optimiztion.

File: server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java
Comment (gashutos) [reply server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java]: Yes. makes sense

File: server/src/main/java/org/opensearch/search/SearchService.java
Comment (gashutos) [reply server/src/main/java/org/opensearch/search/SearchService.java]: This would be too small to cover in UT. I think this will very well covered in integ tests since this boolean setting is enabled by default.

File: server/src/main/java/org/opensearch/index/IndexSettings.java
Comment (reta) [inline server/src/main/java/org/opensearch/index/IndexSettings.java:585]: Minor, just to follow the conventions for on/off features:

```suggestion
        "index.search_segments_reverse_order_optimization.enabled",
```

File: server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java
Comment (reta) [inline server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java:105]:
```suggestion
    private final boolean reverseLeafReaderContexts;
```

**Comment by reta** [comment]
> @nknize @reta , check once you get time, if this appraoch looks good. The optimization gain is pretty good.

I like it @gashutos , a few minor comments but LGTM otherwise, just for the record, it should also work for concurrent search, the segments will be traversed in reverse order concurrently

File: server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java
Comment (gashutos) [reply server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java]: Done

File: server/src/main/java/org/opensearch/index/IndexSettings.java
Comment (gashutos) [reply server/src/main/java/org/opensearch/index/IndexSettings.java]: yeah , makes sense

**Comment by gashutos** [comment]
> 

Awesome, thanks @reta !

> it should also work for concurrent search, the segments will be traversed in reverse order concurrently

I think concurrent search will always land on search() methods of IndexSearcher/ContextIndexSearcher with CollectorManager as an argument. This change is in search() method only with Collector as an argument. So yes, concurrent search will not be benefitted with this order. We will have to override slice() method of IndexSearcher in ContextIndexSearcher to modify sequence of leaves search while preparing slices. I am thinking more how can we enforce the ordering there, doing little more research on that, but concurrent search is not productized yet due to its agressive trade offs so dropped plan to have leaf ordering there.

But for sure, I am thinking to optimize there in next iteration.

**Review by andrross — APPROVED** [review]

File: CHANGELOG.md
Comment (gashutos) [reply CHANGELOG.md]: Sure.

**Comment by andrross** [comment]
> And for other type of workload, it wont regress or impact since data is randomly distributed across all segments.

This _will_ regress any workload that loads data in reverse order and then does descending searches. That does seem like an edge case, whereas time series workloads are extremely common, so I'm sold that this is a good idea to enable by default. However, is the risk of a regression enough of a concern to not change the default behavior in a minor version and only change the default in the next major version?

**Comment by gashutos** [comment]
> > And for other type of workload, it wont regress or impact since data is randomly distributed across all segments.
> 
> This _will_ regress any workload that loads data in reverse order and then does descending searches. That does seem like an edge case, whereas time series workloads are extremely common, so I'm sold that this is a good idea to enable by default. However, is the risk of a regression enough of a concern to not change the default behavior in a minor version and only change the default in the next major version?

Agree. I was thinking in that direction, but later thought would any production customer do that ? Like ingesting ever decreasing workload ? In general, workload can be ever increasing index field or randomly distributed across segments. Even if we use index sort, the sorted order of documents are only within segments, and not really on individual segment order. This will benefit ever increasing index sorted (desc) workload too !.
If there is any workload which is decreasing by index field, that got accidental advatnage and those are rare case, that's why once I thogutht of hardcoding sort field to @timestamp for this optimization and remove this index level enabling/disabling setting. But later realized, workload like nyc_taxy (which is not time series), but if we sort by dropoff_time or pickup_time, that sort is too benefiting 3x (I tested). So thoguht if any such edge case get regressed and CX is really worrying about that, they can disable the flag.
Let me know your thought on it though :) thanks for discussing this....

Another approach I was thinking, if we can sort this leves by its pointValues (maxPointValues) in desc order, that will give us surity that reverse ingested workload wont regress since it will end up resutling in same leaves order. But that will require us to know exact type of sort field so that we can have comparator for that type, (or we can do more generic code to handle that). That is definitly doable but I wasnt simply seeing any worth on that.

@andrross @nknize @reta your input on this would be helpful. That's the crux of this entire approach

**Comment by reta** [comment]
> @andrross @nknize @reta your input on this would be helpful. That's the crux of this entire approach

AFAIR we concluded that this is not a win/win solution for all possible workloads. I think that the time-based workloads are more widely and speeding these ones for a cost of possibly regressing rare edge cases in beneficial. I've just realized that we don't consult the type of sort field but only sort order:

```
request.source().sorts().get(0).order() == SortOrder.DESC
```

May be we should take one more precaution and ensure this is actually DATE/timestamp field?

**Review by nknize — CHANGES_REQUESTED** [review]
> May be we should take one more precaution and ensure this is actually DATE/timestamp field?

💯 I actually think this is very trappy hacky and don't think we should add this the way it is implemented here. 

Here's why: The only common usecase for a presorted field is timeseries and I think this makes sense as the default for that use case only. Any other presorted use case is esoteric. So instead of doing it this way why don't we just do this for timeseries data by adding a segment sorter that is passed to `IndexWriter` through the `InternalEngine` that will sort the segments in reverse order for any timeseries index (e.g., index containing the default `@timestamp` field like is done for DataStreams).

Here's how I would do this:

Since Datastreams is the recommended mechanism for timeseries data, its default time field is [`@timestamp`](https://github.com/opensearch-project/OpenSearch/blob/115bb3072717ba33eb4a35a482e3a7fe10ba354e/server/src/main/java/org/opensearch/index/mapper/DataStreamFieldMapper.java#L42). So make a `reverse` leaf sorter `Comparator` and set it as the default for any index that has the `@timestamp` field defined. 

1. Add a `public static Comparator<LeafReader> TIME_SERIES_LEAF_SORTER... .reverse()`. Maybe to the DataStream class since it's for timeseries data? 

2. You can check if `@timestamp` is defined by adding something like a static utility method `containsTimeStampField` with the logic: `fieldTypeLookup.get(TIMESTAMP_FIELD_NAME);` to  [`MappingLookup`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/mapper/MappingLookup.java)

3. And in IndexShard add a boolean check: `this.isTimeSeriesIndex = mapperService == null ? false : mapperService.mappingLookup().containsTimeStampField();` which is used in `newEngineConfig` (e.g., `new EngineConfig(... isTimeSeriesIndex ? DataStream.TIME_SERIES_LEAF_SORTER : null);`)

4. Add a `Comparator<LeafReader> leafSorter` member to [`EngineConfig#Builder`](https://github.com/opensearch-project/OpenSearch/blob/115bb3072717ba33eb4a35a482e3a7fe10ba354e/server/src/main/java/org/opensearch/index/engine/EngineConfig.java#L459)

5.   [`InternalEngine#getIndexWriterConfig` line 2325](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L2325) add:

```
if (config().getLeafSorter() != null) {
   iwc.setLeafSorter(engineConfig.getLeafSorter());
}
```

This will tell lucene to sort the segments by the comparator passed in. Done!

I think this isolates the optimization to time series, makes it the default, and avoids the trappiness of exposing this to the end user.

**Comment by gashutos** [comment]
> 

Thanks @nknize This was my original implementation too, but the caveat here is, it will regress ASC order query performance (because now we have only one directional traverse of segments since we dont initialize IndexReader again on each query). Like below were my numbers with these implementation on http_logs,

| Sort Type       | hits = 100 |
| ----------- | ----------- | 
| ASC      | 10       | 
| ASC (with optimization)   | 61        | 
| Desc   | 462        | 
| Desc (with optimization)   | 112        | 

Although I agree ASC order on timeseries might not be not that common. (?)

I agree with above approach as it as safe and clean.  But how about below approach.
Add @timestamp field check here in this snippet on query time so it doesn't alter segment reading order for ASC query.

```
//Check if sortField is @timestamp in below logic and remove index setting (since we are sure this will work only on //@timestamp field where we will see optimizations)

  if (this.indexShard.isTimeSeriesIndex()) {
            // Only reverse order for desc order sort queries
            if (request != null
                && request.source() != null
                && request.source().sorts() != null
                && request.source().sorts().size() > 0
                && request.source().sorts().get(0).order() == SortOrder.DESC) {
                return true;
            }
        }
        return false;
    }

```

**Comment by reta** [comment]
> Thanks @nknize This was my original implementation too, but the caveat here is, it will regress ASC order query performance (because now we have only one directional traverse of segments since we dont initialize IndexReader again on each query).

I really like the direction @nknize drives this change, @gashutos do you think it is possible to expand on that and use `LeafSorterFactory` instead of just predefined single `LeafSorter`. The factory could construct the  `LeafSorter` based on sort fields and index type (datastream). I am not sure how hard it is to pass all the context through, the latest improvements certainly looks good, thank you.

File: server/src/main/java/org/opensearch/search/DefaultSearchContext.java
Comment (gashutos) [reply server/src/main/java/org/opensearch/search/DefaultSearchContext.java]: Done

> Let's implement both? Default on timeseries indexes should be descending so lets use the `IndexWriterConfig.setLeafSorter` as the default and only reverse on the infrequent `ASC` query.

Sure. I think that makes sense. Search on laest data set is always preferable on timeseries based workload and I agree making leafSorter in DESC order by default may benefit to queries other than ```sort``` too. 

Let me know @nknize if this looks good, once you approve, I will change (<EDIT>) CR description with EDIT tag that what finally we have merged for the context for others.

**Comment by gashutos** [comment]
> Let's implement both? Default on timeseries indexes should be descending so lets use the `IndexWriterConfig.setLeafSorter` as the default and only reverse on the infrequent `ASC` query.

Let me know @nknize if this looks good, once you approve, I will change (<EDIT>) CR description with EDIT tag that what finally we have merged for the context for others.

File: server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java
Comment (nknize) [inline server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java:114]: Lets add a new ctor so `reverseLeafReaderContexts` isn't a required boolean. That'll minimize the test changes above.

```
    public ContextIndexSearcher(
        IndexReader reader,
        Similarity similarity,
        QueryCache queryCache,
        QueryCachingPolicy queryCachingPolicy,
        boolean wrapWithExitableDirectoryReader,
        Executor executor) {
    this(reader, similarity, queryCache, queryCachingPolicy, wrapWithExitableDirectoryReader, executor, false);
}
```

**Review by nknize — CHANGES_REQUESTED** [review]
Quick skim it's looking better. lets clean up the ctor.

File: server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java
Comment (gashutos) [reply server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java:134]: done.

**Comment by gashutos** [comment]
>do you think it is possible to expand on that and use LeafSorterFactory instead of just predefined single LeafSorter. The factory could construct the LeafSorter based on sort fields and index type (datastream). I am not sure how hard it is to pass all the context through, the latest improvements certainly looks good, thank you.

@reta I was really inclined towards idea of implementing leafSorterFactory so that we can provide leafSorter at both places through that ([InternalEngine](https://github.com/gashutos/OpenSearch/blob/main-timeseries/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L2325) & ContextIndexSearcher). but thought boolean is doing simple job there an we really dont need to sort segments again there, just traverse in reverse. 
Definitly if we find more use-case in future about sorting/sequencing of segments, we will have an interface for ContextIndexSearcher to accept leafSorter at search query level. In face we can have something similar to slice() implementation of concurrent search where each use case can provide their own implementation. 
But as of now lets go with this boolean. Let me know if you think otherwise.

**Comment by nknize** [comment]
> ...we will have an interface for ContextIndexSearcher to accept leafSorter at search query level. In fact we can have something similar to slice() implementation of concurrent search where each use case can provide their own implementation.

I think this is the best way to go for sure. I don't think we have to wait for more use cases as I think there's already enough but I'm happy to see it in a follow up PR and not hold up the simple boolean approach.

File: server/src/main/java/org/opensearch/index/mapper/MappingLookup.java
Comment (reta) [inline server/src/main/java/org/opensearch/index/mapper/MappingLookup.java:269]: Super nit, may is `hasTimeseries()` would be a bit clearer? But no strong argument it is better than current name

Comment (gashutos) [reply server/src/main/java/org/opensearch/index/mapper/MappingLookup.java:269]: ```containsTimeStampField``` this can clearly suggest if index has @timestmp field.
A non-timeseries based index can have too @timestamo field ? (I think yes, and that's where this name fits more)

**Comment by gashutos** [comment]
@nknize are you good now ?

**Review by nknize — APPROVED** [review]
Ship it! Nice and clean! Thx!

**Comment by gashutos** [comment]
@nknize can we add this change to backport 2.x ?

**Pull request merged**
